module/userExtension/helper.js

```diff
@@ -211,304 +212,373 @@ module.exports = class UserExtensionHelper {
-            isAPlatformRole: 1,
-          },
+        for (const csvRow of userRolesCSVData) {
+          const userRole = gen.utils.valueParser(csvRow);
+
+          if (userRole.programs && userRole.programs.length > 0) {
+            userRole.programs.forEach((programId) => allProgramIds.add(programId));
+          }
+
+          if (userRole.user) {
+            allUserIds.add(userRole.user);
+          }
+        }
+
+        // Fetch program data
+        const allProgramsData = await programsHelper.list(
+          { externalId: { $in: Array.from(allProgramIds) } },
+          ['_id', 'externalId'],
+          '',
+          '',
+          '',
+          tenantAndOrgInfo
         );
 
-        let userRoleMap = {};
-        let userRoleAllowedEntityTypes = {};
-
-        userRolesArray.forEach((userRole) => {
-          userRoleMap[userRole.code] = {
-            roleId: userRole._id,
-            code: userRole.code,
-            entities: [],
-            isAPlatformRole: userRole.isAPlatformRole,
-          };
-          userRoleAllowedEntityTypes[userRole.code] = new Array();
-          if (userRole.entityTypes && userRole.entityTypes.length > 0) {
-            userRole.entityTypes.forEach((entityType) => {
-              userRoleAllowedEntityTypes[userRole.code].push(entityType.entityTypeId);
-            });
+        const programIdMap = {};
+        const programInfoMap = {}
+        for (const program of allProgramsData.data.data) {
+          programIdMap[program.externalId] = program._id;
+          programInfoMap[program._id.toString()] = program;
+        }
+
+        // Fetch user profiles
+        const userProfileMap = {};
+        const userProfileResults = await Promise.allSettled(
+          Array.from(allUserIds).map((userId) =>
+            userService
+              .fetchProfileById(tenantAndOrgInfo.tenantId, null, userId)
+              .then((result) => ({ userId, success: true, data: result.data }))
+              .catch((err) => ({ userId, success: false, error: err }))
+          )
+        );
+
+        for (const result of userProfileResults) {
+          if (result.status === 'fulfilled' && result.value.success) {
+            userProfileMap[result.value.userId] = result.value.data;
           }
-        });
+        }
 
-        let userToKeycloakIdMap = {};
-        let userKeycloakId = '';
-        // let userRole;
-        // let existingEntity;
-        // let existingUserRole;
-        const keycloakUserIdIsMandatoryInFile =
-          process.env.DISABLE_LEARNER_SERVICE_ON_OFF && process.env.DISABLE_LEARNER_SERVICE_ON_OFF == 'ON'
-            ? 'true'
-            : false;
+        if(Object.keys(userProfileMap).length === 0) {
+          throw {
+            status: httpStatusCode.bad_request.status,
+            message: messageConstants.apiResponses.USER_NOT_FOUND,
+          };
+        }
+
+        // Fetch user extensions
+        const userExtensionDocs = await database.models.userExtension
```

Comment:
@borkarsaish65 Why are we directly calling db creation/delete/update/read from helper? Don't we have a directory called databaseQueries and from there we channel all db interactions right? Please check this once.

Reply:
done